lang/step.md

```diff
@@ -39,19 +39,35 @@ impl Machine {
 This section defines the following function:
 
 ```rust
-impl Machine {
-    fn eval_value(&mut self, val: ValueExpr) -> NdResult<Value>;
+impl<M: Memory> Machine<M> {
+    fn eval_value(&mut self, val: ValueExpr) -> NdResult<Value<M>>;
 }
 ```
 
 ### Constants
 
-Constants are trivial, as one would hope.
-
 ```rust
-impl Machine {
-    fn eval_value(&mut self, ValueExpr::Constant(value, _ty): ValueExpr) -> NdResult<Value> {
-        value
+impl<M: Memory> Machine<M> {
+    fn eval_constant(&mut self, constant: Constant) -> NdResult<Value<M>> {
+        match constant {
+            Constant::Int(i) => Value::Int(i),
+            Constant::Bool(b) => Value::Bool(b),
+            Constant::Tuple(args) => {
+                let args = args.into_iter()
+                    .map(|arg| self.eval_constant(arg))
+                    .collect();
+                Value::Tuple(args)
+            },
+            Constant::Variant { idx, data } => {
+                let data = self.eval_constant(data);
+                Value::Variant { idx, data }
+            },
+            Constant::Ptr(..) | Constant::Union(..) => panic!("invalid constant encountered!"),
```

Comment:
Please add a comment in well-formed noting that if we allow these consts, we have to also support them here.